### PR TITLE
Unet3d 3plus

### PR DESCRIPTION
**Pull request by aashokvardhan**

### CODE REVIEW

**Comment by vbvg2008** [comment]
1. we need to put the dataset file under fe.dataset.data, then add its download link to the dataset PR test 
2. need to add apphub test script
3. need to modify the apphub readme

**Comment by vbvg2008** [comment]
need some doctrings explaining the dataset

File: apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        encoded_label = np.zeros(list(data.shape) + [self.no_of_classes])
+        for i in range(self.no_of_classes):
+            encoded_label[:, :, :, i] = (data == i).astype(np.uint8)
+        return np.uint8(encoded_label)
+
+
+def get_estimator(epochs=20,
+                  batch_size=1,
+                  input_shape=(256, 256, 24),
+                  channels=(1, ),
```
Comment (vbvg2008) [inline apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py:184]: channels = 1,

File: apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            Sometimes(numpy_op=Rotate(
+                image_in="image", mask_in="label", limit=(-10, 10), border_mode=cv2.BORDER_CONSTANT, mode='train')),
+            ClassEncoding(inputs="label", outputs="label", no_of_classes=num_classes),
+            Minmax(inputs="image", outputs="image"),
+            Minmax(inputs="label", outputs="label", mode='!infer'),
+            ExpandDims(inputs="image", outputs="image"),
+        ])
+
+    # step 2
+    model = fe.build(model_fn=lambda: unet3d_3plus(input_shape + channels, num_classes, filters),
```
Comment (vbvg2008) [inline apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py:213]: input_shape+(channels,)

File: apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    x1 = Conv3D(n_filters, kernel_size=1)(encoder_input)
+    x1 = BatchNormalization()(x1)
+    att = ReLU()(x1 + c1)
+    att = Conv3D(1, kernel_size=1)(att)
+    att = BatchNormalization()(att)
+    att = sigmoid(att)
+    return encoder_input * att
+
+
+class ClassEncoding(NumpyOp):
```
Comment (vbvg2008) [inline apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py:156]: Don't we have our own Onehot numpyOp?

File: apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        eval_data=eval_data,
+        batch_size=batch_size,
+        ops=[
+            Sometimes(numpy_op=HorizontalFlip(image_in="image", mask_in="label", mode='train')),
+            Sometimes(numpy_op=VerticalFlip(image_in="image", mask_in="label", mode='train')),
+            Sometimes(numpy_op=Rotate(
+                image_in="image", mask_in="label", limit=(-10, 10), border_mode=cv2.BORDER_CONSTANT, mode='train')),
+            ClassEncoding(inputs="label", outputs="label", no_of_classes=num_classes),
+            Minmax(inputs="image", outputs="image"),
+            Minmax(inputs="label", outputs="label", mode='!infer'),
```
Comment (vbvg2008) [inline apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py:208]: After onehot encoding, do we still need minmax?  it's either 0 or 1 right?

File: apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        CrossEntropy(inputs=("pred_segment", "label"), outputs="ce_loss", form="binary"),
+        UpdateOp(model=model, loss_name="ce_loss")
+    ])
+
+    # step 3
+    traces = [
+        Dice(true_key="label", pred_key="pred_segment"),
+        ReduceLROnPlateau(model=model, metric="Dice", patience=7, factor=0.5, best_mode="max"),
+        BestModelSaver(model=model, save_dir=save_dir, metric='Dice', save_best_mode='max'),
+        EarlyStopping(monitor="Dice", compare='max', min_delta=0.005, patience=20),
```
Comment (vbvg2008) [inline apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py:230]: epochs=20, and patiences=20 too?  then patience won't do anything.

File: apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
+# Copyright 2019 The FastEstimator Authors. All Rights Reserved.
```
Comment (vbvg2008) [inline apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py:1]: 2022

File: test/apphub_scripts/semantic_segmentation/unet3d_3plus/run_tf.sh [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+set -e
+export TF_CPP_MIN_LOG_LEVEL=2
+
+example_name="unet3d_3plus"
+
+# The training arguments
+# 1. Usually we set the epochs:2, batch_size:2, train_steps_per_epoch:2
+# 2. The expression for the following setup is "--epochs 2 --batch_size 2 --train_steps_per_epoch 2"
+# 3. The syntax of this expression is different from run_notebook.py
+train_info=(--epochs 2 --batch_size 4 --train_steps_per_epoch 2 --eval_steps_per_epoch 2 --input_shape \(32, 32, 16\))
```
Comment (vbvg2008) [inline test/apphub_scripts/semantic_segmentation/unet3d_3plus/run_tf.sh:25]: is "\" necessary?

**Review by vbvg2008 — COMMENTED** [review]
Some changes in tf apply to both torch and notebook

File: apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py (hunk 3 above)
Comment (aashokvardhan) [reply apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py:156]: Our current onehot numpyop supports only 1d array.

File: apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        Resize3D(inputs="label", outputs="label", output_shape=input_shape, mode='!infer'),
+        ModelOp(inputs="image", model=model, outputs="pred_segment"),
+        CrossEntropy(inputs=("pred_segment", "label"), outputs="ce_loss", form="binary"),
+        UpdateOp(model=model, loss_name="ce_loss")
+    ])
+
+    # step 3
+    traces = [
+        Dice(true_key="label", pred_key="pred_segment"),
+        ReduceLROnPlateau(model=model, metric="Dice", patience=7, factor=0.5, best_mode="max"),
```
Comment (vbvg2008) [inline apphub/semantic_segmentation/unet3d_3plus/unet3d_3plus_tf.py:227]: this patience should be less than the early stopping

**Review by vbvg2008 — APPROVED** [review]

**Pull request merged**